sdk/typespec/typespec_client_core/src/fmt.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! Formatting helpers.
+
+use std::borrow::Cow;
+
+/// Converts ASCII characters in `value` to lowercase if required; otherwise, returns the original slice.
+///
+/// # Examples
+///
+/// Returns the original slice:
+///
+/// ```
+/// # use std::borrow::Cow;
+/// # use typespec_client_core::fmt::to_ascii_lowercase;
+/// let actual = to_ascii_lowercase("hello, world!");
+/// assert!(matches!(actual, Cow::Borrowed("hello, world!")));
+/// ```
+///
+/// Returns a clone converted to lowercase ASCII character.
+///
+/// ```
+/// # use std::borrow::Cow;
+/// # use typespec_client_core::fmt::to_ascii_lowercase;
+/// let actual = to_ascii_lowercase("hello, World!");
+/// assert!(matches!(
+///     actual,
+///     Cow::Owned(expected) if expected == "hello, world!"
+/// ));
+/// ```
+pub fn to_ascii_lowercase(value: &str) -> Cow<'_, str> {
+    for (i, c) in value.chars().enumerate() {
+        if c.is_ascii_uppercase() {
+            let mut s = value.to_owned();
+            s[i..].make_ascii_lowercase();
+
+            return Cow::Owned(s);
+        }
+    }
```

Comment:
This feels like a case where the savings over just calling `to_lowercase()` isn't much, given that you have to scan the string anyway, but I'm fine with it if there's a reason for it.

Reply:
Figured no reason to rescan the whole string, or did you mean the purpose for this call anyway? That's fair. My plan was to use it (and I realized I'm not yet, so maybe I should just take it out for now) to normalize header names as I compare against a safe list. My main concern wasn't speed, though, it was allocations. Because we require - at build time - all headers to be lowercase and are encouraging Azure services to make them lowercase, my hope is that most headers we'd scan (think sanitization) are already lowercase and there'd be no need to allocate a new `String`.

/cc @JeffreyRichter

Reply:
Also, that's why this is ASCII-specific: HTTP headers are limited to ASCII. Any non-ASCII character we'll leave as-is and let code elsewhere err or not as appropriate. But since HTTP headers are ASCII, that shouldn't be a practical concern.